crmprtd/align.py

```diff
@@ -8,12 +8,222 @@
+        return None
+    else:
+        return unit_db_check(unit_obs, unit_db, val)
+
+
+def match_station_with_active(sesh, obs_tuple, histories):
+    log.debug('Search for active stations')
+    for history in histories:
+        hist = sesh.query(History).filter(
+            and_(History.id == history.id,
+                 History.sdate != None,
+                 History.edate == None))    # noqa
+
+        if hist.count() == 1:
+            log.debug('Matched history',
+                      extra={'station_name': hist.first().station_name})
+            return hist.first()
+
+    log.warning('Could not find match',
+                extra={'network_name': obs_tuple.network_name,
+                       'native_id': obs_tuple.station_id})
+    return None
+
+
+def match_station_with_location(sesh, obs_tuple, histories):
+    log.debug('Find matching station with location')
+    close_stns = closest_stns_within_threshold(sesh, obs_tuple.lon,
+                                               obs_tuple.lat, 800)
+
+    if len(close_stns) == 0:
+        log.debug('No station nearby')
+        return create_station_and_history_entry(sesh, obs_tuple)
+
+    for id in close_stns:
+        for history in histories:
+            if id == history.id:
+                log.debug('Matched history',
+                          extra={'station_name': history.station_name})
+                return history
+
+
+def match_station(sesh, obs_tuple, history):
+    if obs_tuple.lat and obs_tuple.lon:
+        return match_station_with_location(sesh, obs_tuple, history)
+    else:
+        return match_station_with_active(sesh, obs_tuple, history)
+
+
+def create_station_and_history_entry(sesh, obs_tuple):
+    network = sesh.query(Network).filter(
+        Network.name == obs_tuple.network_name)
+
+    network = network.first()
+    stn = Station(native_id=obs_tuple.station_id, network=network)
+    with sesh.begin_nested():
+        sesh.add(stn)
+    log.info('Created new station entry',
+             extra={'native_id': stn.native_id, 'network_name': network.name})
+
+    hist = History(station=stn,
+                   lat=obs_tuple.lat,
+                   lon=obs_tuple.lon)
+
+    with sesh.begin_nested():
+        sesh.add(hist)
+    log.info('Created new history entry',
+             extra={'history': hist.id, 'network_name': obs_tuple.network_name,
+                    'lat': obs_tuple.lat, 'lon': obs_tuple.lon})
+    sesh.commit()
+    return hist
+
+
+def get_variable(sesh, network_name, variable_name):
+    log.debug('Get variable from db', extra={'var_name': variable_name,
+                                             'network_name': network_name})
+    variable = sesh.query(Variable).join(Network).filter(and_(
+        Network.name == network_name,
+        Variable.name == variable_name)).first()
+
+    if not variable:
+        log.debug('Unable to match variable')
+        return None
+
+    return variable
+
+
+def get_history(sesh, obs_tuple):
```

Comment:
Good, I like this function. It clearly breaks up the various cases (no matches, an exact match, multiple matches) and delegates as necessary. It's clear what's happening.

One improvement that I would suggest is to only pass in the information that is necessary (i.e. the location related info) from the obs_tuple. It will lengthen the argument list, which isn't always desirable, but OTOH it will make it more clear what information the function needs. *And* many of the delegates need a different set of information, so it will make it easier to split up the info.

```
def get_history(sesh, network_name, native_id, lat, lon, elevation):
   ...
```
Would that cover it?

Reply:
When you say ```elevation``` do you mean the threshold used in ```closest_stns_within_threshold()```?

Reply:
No, the vertical elevation of the station. Maybe we don't get that from any of the networks?

Reply:
I haven't see that from any of them.

Reply:
OK, ignore that then. Make sure that there's a big fat warning in the station creation code though, because a station without an elevation is problematic for any further data analysis (particularly because we use elevation-based regression for interpolation).

Reply:
Sounds good, I'll add that in.